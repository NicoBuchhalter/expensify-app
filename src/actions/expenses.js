import uuid from 'uuid';
import firestore from '../firebase/firebase';

const firestoreExpenses = firestore.collection('expenses')

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
}); 

export const startAddExpense = (expensesData = {}) => {
  return (dispatch) => {
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0       
    } =  expensesData;
    const expense = { description, note, amount, createdAt };
    return firestoreExpenses.add(expense).then((docRef) => {
      dispatch(addExpense({
        id: docRef.id, 
        ...expense
      }));
    });

  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return firestoreExpenses.doc(id).delete().then(() => {
      dispatch(removeExpense({ id }));
    });    
  }
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});


export const startSetExpenses = () => {
  return (dispatch) => {
    return firestoreExpenses.get().then(querySnapshot => {
      const expenses = querySnapshot.docs.map(docRef => ({
        id: docRef.id,
        ...docRef.data()
      }));
      dispatch(setExpenses(expenses));
    });    
  }
};