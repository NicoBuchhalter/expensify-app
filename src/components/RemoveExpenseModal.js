import React from 'react';
import Modal from 'react-modal';


const RemoveExpenseModal = (props) => (
  <Modal 
    isOpen={props.isModalOpen} 
    contentLabel="Remove Expense"
    closeTimeoutMS={200}
    className="modal"
  >
    <h3 className='modal__title'>{props.expenseDescription}</h3>
    <p className='modal__body'>Are you sure you want to remove this expense?</p>
    <div className='modal__buttons'>
      <button className='button button--secondary' onClick={props.handleRemoveExpense}>Remove</button>
      <button className='button' onClick={props.handleCloseModal}>Cancel</button>
    </div>
  </Modal>
)

export default RemoveExpenseModal;