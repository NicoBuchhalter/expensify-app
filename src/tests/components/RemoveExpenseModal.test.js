import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import RemoveExpenseModal from '../../components/RemoveExpenseModal';

let handleRemoveExpense, handleCloseModal, history, wrapper;
beforeEach(() => {
  handleRemoveExpense = jest.fn();
  handleCloseModal = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <RemoveExpenseModal
      isModalOpen={true}
      expenseDescription={expenses[1].description} 
      handleRemoveExpense={handleRemoveExpense}
      handleCloseModal={handleCloseModal} 
      history={history} 
    />
  );
});

test('Should render RemoveExpenseModal correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('Should handle remove', () => {
  wrapper.find('.button--secondary').simulate('click');
  expect(handleRemoveExpense).toHaveBeenCalled();
});

test('Should handle cancel', () => {
  wrapper.find('button').last().simulate('click');
  expect(handleCloseModal).toHaveBeenCalled();
});