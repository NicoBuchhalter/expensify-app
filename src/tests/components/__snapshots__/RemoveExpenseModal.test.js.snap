// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render RemoveExpenseModal correctly 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={200}
  contentLabel="Remove Expense"
  isOpen={true}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <h3
    className="modal__title"
  >
    Rent
  </h3>
  <p
    className="modal__body"
  >
    Are you sure you want to remove this expense?
  </p>
  <div
    className="modal__buttons"
  >
    <button
      className="button button--secondary"
      onClick={[MockFunction]}
    >
      Remove
    </button>
    <button
      className="button"
      onClick={[MockFunction]}
    >
      Cancel
    </button>
  </div>
</Modal>
`;
